code/game/objects/items/devices/scanners.dm

```diff
@@ -966,3 +967,65 @@ GENE SCANNER
 	playsound(src, 'sound/machines/ping.ogg', 30, TRUE)
 	return TRUE
 
+
+/obj/item/material_scanner
+	name = "material scanner"
+	desc = "A device that analyzes a object's composition and lists its materials."
+	icon = 'icons/obj/device.dmi'
+	icon_state = "forensic0"
+	item_state = "analyzer"
+	lefthand_file = 'icons/mob/inhands/equipment/tools_lefthand.dmi'
+	righthand_file = 'icons/mob/inhands/equipment/tools_righthand.dmi'
+	w_class = WEIGHT_CLASS_SMALL
+	flags_1 = CONDUCT_1
+	throwforce = 0
+	throw_speed = 3
+	throw_range = 7
+	materials = list(/datum/material/iron=30, /datum/material/glass=20)
+
+/obj/item/material_scanner/afterattack(atom/target, mob/user, proximity_flag, click_parameters)
+	. = ..()
+	if(istype(target,/obj/item))
+		var/obj/item/I = target
+		if (I.materials && length(I.materials)>0)
+			analyze_materials(I, user)
+			return
+	to_chat(user, "<span class='notice'>[target] doesn't have any reusable materials.</span>")
+
+/obj/item/material_scanner/proc/analyze_materials(obj/item/I, mob/living/user)
+	to_chat(user, "- Material Analysis of [I] -")
+	var/integ = 1
+	var/output = ""
+	for(var/R in I.materials)
```

Comment:
This should definitely be an assoc list

Reply:
Assoc list? I don't think I've done those before. Can you give an example?